### PR TITLE
iox-#1196 Fix expected warnings

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
    - [ ] Each unit test case has a unique UUID
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Part of #1196

---
This change is [Reviewable](https://reviewable.io/reviews/apexai/iceoryx/24)

### CODE REVIEW

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/expected.inl [hunk 1]
```diff
@@ -198,6 +198,8 @@ inline ValueType* expected<ValueType, ErrorType>::operator->() noexcept
 template <typename ValueType, typename ErrorType>
 inline const ValueType* expected<ValueType, ErrorType>::operator->() const noexcept
 {
+    // const_cast avoids code duplication
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/expected.inl]: Not sure, if we should also clarify why the const_cast is safe

File: iceoryx_hoofs/test/.clang-tidy [hunk 2]
```diff
@@ -31,3 +31,15 @@ Checks: '
...
+  - { key: readability-function-size.LineThreshold,               value: 10000 }
+  - { key: readability-function-size.BranchThreshold,             value: 10000 }
+  - { key: readability-function-size.StatementThreshold,          value: 10000 }
+  - { key: readability-function-cognitive-complexity.Threshold,   value: 10000 }
+
+## Justifications for CheckOptions ##
+#
+# EXPECT_*, ASSERT_* and TEST_F, TYPED_TEST macros are introducing additional
+# branches and increases the code complexity without affecting the user. Therefore,
+# those metrics are hardly applicable in tests.
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/test/.clang-tidy]:
```suggestion
# EXPECT_*, ASSERT_* and TEST_F, TYPED_TEST macros introduce additional
# branches and increase the code complexity without affecting the user. Therefore,
# those metrics are hardly applicable in tests.
```

File: iceoryx_hoofs/test/moduletests/test_cxx_expected.cpp [hunk 3]
```diff
@@ -35,12 +36,17 @@ class MockCallables
 
 struct TestClass
 {
+    // the parameters can be swapped without changing the outcome, methods which
+    // work on them are commutative
+    // NOLINTNEXTLINE( bugprone-easily-swappable-parameters)
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/test/moduletests/test_cxx_expected.cpp]:
```suggestion
    // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
```

File: iceoryx_hoofs/test/moduletests/test_cxx_expected.cpp [hunk 4]
```diff
@@ -62,6 +68,8 @@ struct TestClass
 
 struct NonTrivialTestClass
 {
+    // swapped parameters will be directly detected with failing tests
+    // NOLINTNEXTLINE( bugprone-easily-swappable-parameters)
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/test/moduletests/test_cxx_expected.cpp]:
```suggestion
    // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
```

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/expected.hpp [hunk 5]
```diff
@@ -403,13 +431,11 @@ class IOX_NO_DISCARD expected<ValueType, ErrorType>
...
     /// @endcode
+    //
     template <typename T>
+    // NOLINTNEXTLINE(hicpp-explicit-conversions) see doxygen brief section
     operator expected<T>() const noexcept;
 
-    /// @brief conversion operator to an optional.
-    /// @return optional containing the value if the expected contains a value, otherwise a nullopt
-    operator optional<ValueType>() const noexcept;
-
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/expected.hpp:412]: Do we need to add this to the release notes as breaking change? ... I think I have to do the same for the Duration operator timeval

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/expected.inl [hunk 6]
```diff
@@ -235,22 +239,16 @@ inline optional<ValueType> expected<ValueType, ErrorType>::to_optional() const n
...
 inline expected<ErrorType>::expected(variant<ErrorType>&& f_store) noexcept
     : m_store(std::move(f_store))
 {
 }
 
 template <typename ErrorType>
-inline expected<ErrorType>::expected(const success<void>&) noexcept
+inline expected<ErrorType>::expected(const success<void>& successValue) noexcept
 {
+    IOX_DISCARD_RESULT(successValue);
```
Comment (elBoberido) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/expected.inl]:
```suggestion
inline expected<ErrorType>::expected(const success<void>& IOX_MAYBE_UNUSED(successValue)) noexcept
{
```
The code above should also work and is semantically more correct

**Review by elBoberido — APPROVED** [review]

**Pull request merged**